specifyweb/stored_queries/format.py

```diff
@@ -89,96 +101,125 @@ def pseudo_sprintf(self, format, expr):
-            if specify_field.is_relationship:
-                formatter_name = fieldNode.attrib.get('formatter', None)
-                query, expr = self.objformat(query, table, formatter_name)
             else:
-                expr = self._fieldformat(specify_field, getattr(table, specify_field.name))
+                query, table, model, specify_field = query.build_join(
+                    specify_model, orm_table, formatter_field_spec.join_path)
+                expr = self._fieldformat(formatter_field_spec.get_field(),
+                                         getattr(table, specify_field.name))
+
+
+            if 'format' in fieldNodeAttrib:
+                expr = self.pseudo_sprintf(fieldNodeAttrib['format'], expr)
 
-            if 'format' in fieldNode.attrib:
-                expr = self.pseudo_sprintf(fieldNode.attrib['format'], expr)
 
-            if 'sep' in fieldNode.attrib:
-                expr = concat(fieldNode.attrib['sep'], expr)
-            return query, blank_nulls(expr)
+            if 'sep' in fieldNodeAttrib:
+                expr = concat(fieldNodeAttrib['sep'], expr)
+            return query, blank_nulls(expr), formatter_field_spec
 
-        def make_case(query: QueryConstruct, caseNode: Element) -> Tuple[QueryConstruct, Optional[str], blank_nulls]:
+        def make_case(query: QueryConstruct, caseNode: Element) -> Tuple[
+            QueryConstruct, Optional[str], blank_nulls]:
             field_exprs = []
             for node in caseNode.findall('field'):
-                query, expr = make_expr(query, node)
+                query, expr, _ = make_expr(query, node.text, node.attrib)
                 field_exprs.append(expr)
 
             expr = concat(*field_exprs) if len(field_exprs) > 1 else field_exprs[0]
-            return query, case_value_convert(caseNode.attrib.get('value', None)), expr
+            return query, caseNode.attrib.get('value', None), expr
 
+        switchNode = formatterNode.find('switch')
+        single = switchNode.attrib.get('single', 'true') == 'true'
         cases = []
         for caseNode in switchNode.findall('fields'):
             query, value, expr = make_case(query, caseNode)
             cases.append((value, expr))
 
         if not cases:
-            logger.warn("dataobjformatter for %s contains switch clause no fields", specify_model)
+            logger.warn(
+                "dataobjformatter for %s contains switch clause no fields",
+                specify_model)
             return query, literal(_("<Formatter not defined.>"))
 
         if single:
             value, expr = cases[0]
         else:
-            control_field = getattr(orm_table, switchNode.attrib['field'])
-            expr = case(cases, control_field)
-
+            query, formatted, swith_field_spec = make_expr(query, switchNode.attrib['field'], {})
+            def case_value_convert(value): return value == 'true' if swith_field_spec.get_field().type == 'java.lang.Boolean' else value
```

Comment:
huh, didn't know python allowed that! I though it required return to be on a new line

Reply:
Just some weird things Python allows you to shoot your foot with